### ISSUE TITLE
Consider putting math kernels into standalone module?

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs#106]
_From @ofrobots on January 3, 2018 21:53_

Have we considered extracting the math functionality from this module and putting it in a standalone module by itself. Math functionality seems independently useful.

_Copied from original issue: tensorflow/tfjs-core#508_

**Comment by nsthorat** [comment]
_From @dsmilkov on January 4, 2018 0:50_

Hi, to better understand, can you share your use case? `NDArrayMath` doesn't make sense  without `NDArray`s. Did you mean `NDArrayMath` independent of the graph/gradient/backpropagation code?

**Comment by nsthorat** [comment]
_From @ofrobots on January 4, 2018 1:6_

Right, sorry. Yeah, I think we have a really good implementation of a accelerated math library here. I think it can be useful independently of the remainder of the ML code.

**Comment by nsthorat** [comment]
_From @ofrobots on January 22, 2018 18:55_

Some experiences from over this weekend. I was working a backend (Node.js) project where I just needed a 3D Array structure, I started with deeplearn as it has a really nice accelerated implementation. Here's the friction log:

* It would be good to be able to use just the NDArray and associated code without depending upon the entirety of deeplearn. (This issue)
* In my Node.js typescript project, I had to enable `lib: ["dom"]` in `tsconfig.json`. This module directly depends upon browser APIs which makes adds to the friction of using it on the backend. Only the CPU Math backend is available on Node.js presently; and from a JavaScript point of view, things work. From a types point of view they don't. (I can extract this into a standalone issue).
* My typescript project follows the Google TypeScript style guide which uses `strict` compiler flag. The compiler complains of type errors in the deeplearn.js code. The work-around is to disable strict mode which is highly undesirable. (issue #576)
